test/core/xds/xds_client_test.cc

```diff
@@ -3991,48 +4033,125 @@ TEST_F(XdsClientTest, FallbackReportsError) {
+      << error->message();
+  resource = watcher2->WaitForNextResource();
+  ASSERT_NE(resource, nullptr);
+  EXPECT_EQ(resource->name, "foo1");
+  EXPECT_EQ(resource->value, 6);
+  error = watcher2->WaitForNextError();
+  ASSERT_TRUE(error.has_value());
+  EXPECT_THAT(error->message(),
+              ::testing::ContainsRegex("Another server down"));
+  CancelFooWatch(watcher.get(), "foo1");
+  CancelFooWatch(watcher2.get(), "foo1");
+}
+
+TEST_F(XdsClientTest, FallbackOnStartup) {
+  FakeXdsBootstrap::FakeXdsServer primary_server;
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  // Regular operation
+  InitXdsClient(FakeXdsBootstrap::Builder().SetServers(
+      {primary_server, fallback_server}));
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
+  TriggerConnectionFailure(primary_server,
+                           absl::UnavailableError("Primary server is down"));
+  EXPECT_THAT(
+      GetServerConnections(),
+      ::testing::ElementsAre(::testing::Pair(kDefaultXdsServerUrl, false),
+                             ::testing::Pair("fallback_xds_server", true)));
+  // XdsClient should have created an ADS stream.
+  auto fallback_stream = WaitForAdsStream(fallback_server);
+  ASSERT_NE(fallback_stream, nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(fallback_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  // Send a response.
+  fallback_stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddFooResource(XdsFooResource("foo1", 6))
+          .Serialize());
+  // XdsClient should have delivered the response to the watcher.
+  auto resource = watcher->WaitForNextResource();
+  ASSERT_NE(resource, nullptr);
+  EXPECT_EQ(resource->name, "foo1");
+  EXPECT_EQ(resource->value, 6);
+  request = WaitForRequest(fallback_stream.get());
   ASSERT_TRUE(request.has_value());
   CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
                /*version_info=*/"1", /*response_nonce=*/"A",
                /*error_detail=*/absl::OkStatus(),
-               /*resource_names=*/{"foo1", "foo2"});
-  auto stream2 = WaitForAdsStream(fallback_server);
-  TriggerConnectionFailure(fallback_server,
-                           absl::UnavailableError("Another server down"));
-  ASSERT_TRUE(stream2 != nullptr);
-  // XdsClient should have sent a subscription request on the ADS stream.
-  request = WaitForRequest(stream2.get());
+               /*resource_names=*/{"foo1"});
+  // Recover to primary
+  TriggerConnectionFailure(primary_server, absl::OkStatus());
```

Comment:
As mentioned above, you should not call this to restore connectivity.  Instead, just send a response on the stream.